### ISSUE TITLE
Commonjs imports fail Electron App startup

### ISSUE THREAD

**Issue by Calhistorian** [oramasearch/orama#277]
**Describe the bug**
I am building an existing electron application. I have been using Lyra v0.3.1 using commonjs imports.
e.g. `const { create, search, insert, remove, insertBatch } = require('@lyrasearch/lyra');`

However, when trying to use v0.4+ I get an error on startup in the dev console:
<img width="583" alt="CleanShot 2023-02-07 at 22 08 12@2x" src="https://user-images.githubusercontent.com/51142180/217448744-ebdd083e-c286-40be-b2ad-e5e4de0cae97.png">

This completely prevents the app from rendering.

**To Reproduce**
1. Remove v0.3.1 version commonjs import: `const { create, search, insert, remove, insertBatch } = require('@lyrasearch/lyra');`
2. Replace with v0.4+ commonjs import: `const { search, insert, remove, insertBatch, create } = require("@lyrasearch/lyra/cjs");`
3. Start application - no other things have changed

**Expected behavior**
The application should load, and if any error, throw.

**Screenshots**
See above

**Desktop (please complete the following information):**
 - OS: macOS v13.1
 - Browser: n/a (Electron)
 - Version: 21.1.0

**Additional context**
I am not that savvy of a JavaScript developer, and if this has not come up as a known problem it is likely a configuration problem on my end. I just expected to be able to replace the import statement and go on as usual.

**Comment by micheleriva** [comment]
cc. @ShogunPanda

**Comment by ShogunPanda** [comment]
@Calhistorian I'm not familiar with Electron instead. Can you please provide a repro repository so I can debug the issue?

**Comment by Calhistorian** [comment]
As an update, it seems there is a fundamental error that needs to be solved first to get to the bottom of the issue. Both import and require versions of Lyra 4.11, and Orama 4.12 fail to load.

The import version throws a relative path error:
<img width="1442" alt="image" src="https://user-images.githubusercontent.com/51142180/226829157-36993f39-bde1-4039-a2bd-fd0da4a82b78.png">

The require version throws a require not defined error
<img width="1665" alt="image" src="https://user-images.githubusercontent.com/51142180/226829215-37c4a740-2f92-456e-83d9-203990820e12.png">

Note: 
`nodeIntegration = true`
require works fine in the main process

Related Electron Issue that is related to require and import problems:
https://github.com/electron/electron/pull/37535

**Comment by GitMurf** [comment]
@Calhistorian can you try using this import statement instead? You just have to specify the full path via node_modules when imported via the renderer (browser).

```javascript
import { create, insert, remove, search } from './node_modules/@orama/orama/dist/index.js';
```

When I then use the following, I do not get the dev tools crash that you reported. This is on a fresh Electron install NOT using Forge so give this same code a try with the updated import path with your example and see if it works without crashing?

```javascript
// renderer.js
console.log('renderer.js loaded');
import { create, insert, remove, search } from './node_modules/@orama/orama/dist/index.js';

const db = await create({
  schema: {
    author: 'string',
    quote: 'string',
  },
})
console.log(db);
```

Here is a screenshot of the console results showing the `db` created successfully:

![image](https://user-images.githubusercontent.com/64155612/227025201-0aada64b-fc35-406f-9250-4d645623c828.png)

I tried to create it exactly as you with simple code below:

```html
/* index.html */
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8" />
    <title>Hello World!</title>
    <link rel="stylesheet" href="index.css" />
    <script type="module" src="./renderer.js" async defer ></script>
</head>
<body>
    <h1>Hello from Electron renderer!</h1>
    <p>👋</p>
</body>
</html>
```

```javascript
// renderer.js
console.log('renderer.js loaded');
import { create, insert, remove, search } from './node_modules/@orama/orama/dist/index.js';

const db = await create({
  schema: {
    author: 'string',
    quote: 'string',
  },
})
console.log(db);
```

```javascript
// main.js
const { app, BrowserWindow } = require('electron');
const { join } = require('path');

let win;

async function createWindow() {
  win = new BrowserWindow({
    width: 800,
    height: 600,
    webPreferences: {
      nodeIntegration: false, // is default value after Electron v5
      contextIsolation: true, // protect against prototype pollution
      enableRemoteModule: false, // turn off remote
      sandbox: true, // sandbox the renderer for added security
    }
  });
  win.loadFile(join(__dirname, 'index.html'));
}

app.on('ready', () => {
  createWindow();
  app.on('activate', () => {
    if (BrowserWindow.getAllWindows().length === 0) {
      createWindow();
    }
  });
});

app.on('window-all-closed', () => {
  if (process.platform !== 'darwin') {
    app.quit();
  }
});
```

**Comment by GitMurf** [comment]
As an FYI this is using the newest version (as of right now what `npm install @orama/orama` gives you) of Orama: `"@orama/orama": "^1.0.0-beta.4"`

**Comment by Calhistorian** [comment]
Confirmed: the path issues continue. It does work with Electron-Forge with the direct path and ESM.

<img width="1168" alt="CleanShot 2023-03-22 at 13 22 16@2x" src="https://user-images.githubusercontent.com/51142180/227029510-07f17aba-f7ab-479b-a38d-48a86a813338.png">

However Electron fails with the commonJS require:

<img width="1444" alt="CleanShot 2023-03-22 at 13 26 58@2x" src="https://user-images.githubusercontent.com/51142180/227030009-2473cf56-bf03-4c0c-bcb0-269065842cc7.png">

As it stands I will close this issue pending further testing in production and/or Electron solves their "require" issues.

**Comment by GitMurf** [comment]
@Calhistorian require() will not work in the "browser" (renderer) as it is a Node thing. So that is where if we want to require something in the renderer process then we have to turn off a lot of the Electron recommended security protections like nodeIntegration, contextIsolation and sandbox.

Change the web prefs to the following:

```javascript
    webPreferences: {
      nodeIntegration: true, // is default value after Electron v5
      contextIsolation: false, // protect against prototype pollution
      enableRemoteModule: false, // turn off remote
      sandbox: false, // sandbox the renderer for added security
    }
```

And then in the renderer.js use this instead. Notice I wrap the call of using create() in a setTimeout of 7 seconds just so you can see Lyra imported fine but then in 7 seconds the dev tools issue will present itself. So `require()` should work as well with Electron, but now there is an issue when trying to use it (crashes).

```js
const { create, insert, remove, search } = require('@orama/orama');

setTimeout(async () => {
  const db = await create({
    schema: {
      author: 'string',
      quote: 'string',
    },
  })
  console.log(db);
}, 7000);
```

My hunch is the issue has to do with using Node for `require()` but then running Lyra (Orama) in the browser (renderer).

**Comment by GitMurf** [comment]
Ok, yes my hunch I believe is correct! If I take the same require() code and put in the main process instead of renderer, then it works without crashing Electron. So it must be because you are trying to use require()/commonjs (which is likely built specifically for Node maybe?) but using it in the Browser (renderer).

Here is total code with BOTH require() working in Main.js and import() working in renderer without the dev tool crash. Also this has all the security settings turned back on (because not using require() in renderer).

```js
// main.js

const { app, BrowserWindow } = require('electron');
const { join } = require('path');
const lyra = require('@orama/orama');

let win;

async function createWindow() {
  win = new BrowserWindow({
    width: 800,
    height: 600,
    webPreferences: {
      nodeIntegration: false, // is default value after Electron v5
      contextIsolation: true, // protect against prototype pollution
      enableRemoteModule: false, // turn off remote
      sandbox: true, // sandbox the renderer for added security
    }
  });
  win.loadFile(join(__dirname, 'index.html'));
}

app.on('ready', async () => {
  createWindow();
  app.on('activate', () => {
    if (BrowserWindow.getAllWindows().length === 0) {
      createWindow();
    }
  });

  console.log('main.js loaded');
  console.log('Lyra keys:', Object.keys(lyra));
  const db = await lyra.create({
    schema: {
      author: 'string',
      quote: 'string',
    },
  })
  console.log(db);
});

app.on('window-all-closed', () => {
  if (process.platform !== 'darwin') {
    app.quit();
  }
});

```

```js
// renderer.js

console.log('renderer.js loaded');
import { create, insert, remove, search } from './node_modules/@orama/orama/dist/index.js';

const db = await create({
  schema: {
    author: 'string',
    quote: 'string',
  },
})
console.log(db);
```

```html
/* index.html */

<!DOCTYPE html>
<html>

<head>
    <meta charset="UTF-8" />
    <title>Hello World!</title>
    <link rel="stylesheet" href="index.css" />
    <script type="module" src="./renderer.js" async defer ></script>
</head>

<body>
    <h1>Hello from Electron renderer!</h1>
    <p>👋</p>
</body>

</html>
```

**Comment by ShogunPanda** [comment]
Hello!
All the issues above should have been fixed in Orama 1.0.0-beta.6 - Can you please take a look and confirm it?

Note that you don't need to import from `@orama/orama/cjs` and just use `@orama/orama`.

If you run in a error about the stemmer reporting on a link on how to fix the issue, the link is currently wrong and will be fixed in beta 7.
The right link is: https://docs.oramasearch.com/text-analysis/stemming#using-stemming-with-bundlers
